services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy

```diff
@@ -0,0 +1,994 @@
+
+        where:
+        bandwidth | vlanId
+        1000      | 3378
+        0         | 3378
+        1000      | 0
+        0         | 0
+    }
+
+    def "Able to update the 'protected path' option"() {
+        given: "Two active not neighboring switches with two possible paths at least"
+        def (Switch srcSwitch, Switch dstSwitch) = getNotNeighboringSwitchPair(2)
+
+        when: "Create flow without protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = false
+        flowHelper.addFlow(flow)
+
+        then: "Flow is created without protected path"
+        !northbound.getFlowPath(flow.id).protectedPath
+
+        and: "Needed number of rules are created"
+        northbound.getSwitchRules(srcSwitch.dpId).flowEntries.findAll { !Cookie.isDefaultRule(it.cookie) }.size() == 2
+
+        pathHelper.getInvolvedSwitches(flow.id)[1..-1].each { sw ->
+            assert northbound.getSwitchRules(sw.dpId).flowEntries.findAll {
+                !Cookie.isDefaultRule(it.cookie)
+            }.size() == 2
+        }
+
+        when: "Update flow: enable protected path(allocateProtectedPath=true)"
+        def currentLastUpdate = northbound.getFlow(flow.id).lastUpdated
+        northbound.updateFlow(flow.id, flow.tap { it.allocateProtectedPath = true })
+
+        then: "Protected path is enabled"
+        def flowPathInfoAfterUpdating = northbound.getFlowPath(flow.id)
+        flowPathInfoAfterUpdating.protectedPath
+
+        currentLastUpdate < northbound.getFlow(flow.id).lastUpdated
+
+        and: "Rules are updated on the main path"
+        Wrappers.wait(WAIT_OFFSET) {
+            flowHelper.verifyRulesOnProtectedFlow(flow.id)
+        }
+
+        when: "Update flow: disable protected path(allocateProtectedPath=false)"
+        def mainFlowPath = northbound.getFlowPath(flow.id).forwardPath
+        def protectedFlowPath = northbound.getFlowPath(flow.id).protectedPath.forwardPath
+        northbound.updateFlow(flow.id, flow.tap { it.allocateProtectedPath = false })
+
+        then: "Protected path is disabled"
+        !northbound.getFlowPath(flow.id).protectedPath
+
+        and: "Rules for protected path are deleted"
+        def commonNodeIds = mainFlowPath*.switchId.intersect(protectedFlowPath*.switchId)
+        Wrappers.wait(WAIT_OFFSET) {
+            commonNodeIds.each { sw ->
+                def rules = northbound.getSwitchRules(sw).flowEntries.findAll {
+                    !Cookie.isDefaultRule(it.cookie)
+                }
+                if (sw == srcSwitch.dpId || sw == dstSwitch.dpId) {
+                    assert rules.size() == 2
+
+                    def mainNode = mainFlowPath.find { it.switchId == sw }
+                    assert flowHelper.findForwardPathRules(rules, mainNode).size() == 1
+                    assert flowHelper.findReversePathRules(rules, mainNode).size() == 1
+                } else {
+                    assert rules.size() == 4
```

Comment:
It looks like this is the wrong assumption.

Reply:
Fixed
```
       protectedFlowPath.each { sw ->
            def rules = northbound.getSwitchRules(sw.switchId).flowEntries.findAll {
                !Cookie.isDefaultRule(it.cookie)
            }

            assert flowHelper.findForwardPathRules(rules, sw).size() == 0
            assert flowHelper.findReversePathRules(rules, sw).size() == 0
        }
```